### DISCUSSION TITLE
Can I add return types to a custom collection for autocompletion?

### DISCUSSION THREAD

**Discussion by ralphjsmit** [laravel/framework#39679] [Q&A] [answered]
Hey everyone, I've created a custom collection and I'm using that as a DTO. The custom collection is quite simple, it only contains a static `make()` override, because this make override casts all items in the collection as a specific other DTO.

This other DTO is just used for accessing a few properties and methods. One of the big advantages of DTOs over just arrays is that DTO classes provide autocompletion for properties and methods.

Unfortunately my IDE (PhpStorm) isn't able to figure out the return type of things like `->first()`, `->last()` etc, whilst this would be very handy (preventing typos and speeding up the workflow).

I already tried a few things, like adding a `@return` on the whole collection class (in vain😂). Does anyone have an idea how to implement / achieve this?

Personally I believe that this will be handy for many use cases, because most collections only contain one item type.

Many thanks

```php
<?php

namespace App\DTOs;

use Illuminate\Support\Carbon;
use Illuminate\Support\Collection;

class MyDTO extends Collection
{
    public static function make($items = []): static
    {
        return (new static($items))->map(function($input) {
            return new OtherDTO($input);
        });
    }
}

**Comment by PaolaRuby** [comment, marked as answer]
```php
$MyDTOCollection->map(function($input) {
    /** @var OtherDTO $input */
   $input->auto_completion
});
```

Also [laravel-ide-helper](https://github.com/barryvdh/laravel-ide-helper) could help

**Reply by ralphjsmit** [reply]
Thanks for your suggestion! I tried this and it does indeed work, but it would require to add the docblock everywhere, which is not perfect 😊

I checked with laravel-ide-helper, but it looks like they don’t have full support for the collections (yet).

**Reply by PaolaRuby** [reply]
I think that is not posible because var is inside of callable, the best approach is to declare the type in the callable without doc, example:
```php
$MyDTOCollection->each(function(OtherDTO  $input) {
   $input->auto_completion
});
```
And now your code is better using types(Best Practices)

----

Also why create new colecctions? `class MyDTO extends Collection`, is'nt it better just creating a macro?
```php
Collection::macro('makeDTO', function(string $class) {
    return $this->map(function($input) use ($class) {
        return new $class($input);
    });
});
// Model::get()->makeDTO(OtherDTO::class);
// collect($items)->makeDTO(OtherDTO::class)->each(function(OtherDTO  $input) {
```
???

**Reply by ralphjsmit** [reply]
Adding type definitions is a good idea, I'm already doing that for most things so why not callbacks? ;-)

Creating a macro might be nice for situations where there's relatively simply stuff being done, this custom collection has a multiple larger methods. Using a custom collection is a much cleaner.

Thanks!

**Comment by Cellard** [comment]
```php
<?php

namespace App\DTOs;

use Illuminate\Support\Carbon;
use Illuminate\Support\Collection;

/**
 * @method OtherDTO first(callable $callback = null, $default = null)
 */
class MyDTO extends Collection
{
    public static function make($items = []): static
    {
        return (new static($items))->map(function($input) {
            return new OtherDTO($input);
        });
    }
}
```

**Reply by ralphjsmit** [reply]
Thanks for your suggestion! This works quite nice for first(), last(), etc (though you have to add a docblock for every method). The downside is that it doesn’t work inside each() statements. I upvoted your answer, but I’ll wait a week with marking it as correct to see if there might come another answer.👍
